contracts/ColonyStorage.sol

```diff
@@ -125,13 +125,46 @@ contract ColonyStorage is CommonStorage, ColonyDataTypes, DSMath {
     _;
   }
 
-  modifier isAdmin(address _user) {
-    require(ColonyAuthority(address(authority)).hasUserRole(_user, uint8(ColonyRole.Admin)), "colony-not-admin");
+  modifier isAdmin(uint256 _permissionDomainId, uint256 _childSkillIndex, uint256 _id, address _user) {
+    require(ColonyAuthority(address(authority)).hasUserRole(_user, _permissionDomainId, uint8(ColonyRole.Administration)), "colony-not-admin");
+    require(validateDomainProof(_permissionDomainId, _childSkillIndex, tasks[_id].domainId), "colony-invalid-domain-proof");
     _;
   }
 
   modifier self() {
     require(address(this) == msg.sender, "colony-not-self");
     _;
   }
+
+  modifier auth {
+    require(isAuthorized(msg.sender, 1, msg.sig), "ds-auth-unauthorized");
+    _;
+  }
+
+  modifier authDomain(uint256 _permissionDomainId, uint256 _childSkillIndex, uint256 _childDomainId) {
+    require(isAuthorized(msg.sender, _permissionDomainId, msg.sig), "ds-auth-unauthorized");
+    require(validateDomainProof(_permissionDomainId, _childSkillIndex, _childDomainId), "ds-auth-invalid-domain-proof");
+    if (canCallBecauseArchitect(msg.sender, _permissionDomainId, msg.sig)) {
+      require(_permissionDomainId != _childDomainId, "ds-auth-only-authorized-in-child-domain");
+    }
+    _;
+  }
+
+  // Evaluates a "domain proof" which checks that childDomainId is part of the subtree starting at permissionDomainId
+  function validateDomainProof(uint256 permissionDomainId, uint256 childSkillIndex, uint256 childDomainId) internal view returns (bool) {
+    return (permissionDomainId == childDomainId) || (getChildSkillId(permissionDomainId, childSkillIndex) == domains[childDomainId].skillId);
+  }
+
+  function getChildSkillId(uint256 permissionDomainId, uint256 childSkillIndex) internal view returns (uint256) {
+    return IColonyNetwork(colonyNetworkAddress).getChildSkillId(domains[permissionDomainId].skillId, childSkillIndex);
+  }
+
+  function canCallBecauseArchitect(address src, uint256 domainId, bytes4 sig) internal view returns (bool) {
+    return DomainRoles(address(authority)).canCallBecause(src, domainId, uint8(ColonyRole.ArchitectureSubdomain), address(this), sig);
```

Comment:
I haven't been able to explain why this returns `false` for the founder? The founder belonging to the `ArchitectureSubdomain` role is explicitly tested [here](https://github.com/JoinColony/colonyNetwork/pull/553/files#diff-c255652ac9460b92a219642618a4887cR65) yet if you try 
```
const canCallMakeTaskBecauseArchitect = await colonyAuthority.canCallBecause(FOUNDER, 1, ARCHITECTURE_ROLE, colony.address, getFunctionSignature("makeTask(uint256,uint256,bytes32,uint256,uint256,uint256)"));
``` 
in that test it returns `false` 🤔

Reply:
Mm, so this function checks to see if it is _only_ the architecture permission that gives this role. I.e. if you have root and architecture funding then it returns `false`, but if you have only architecture permission it returns `true`. That was how we implemented the restriction on the architecture role where you can only call certain functions in a _subdomain_ of your permission,  i.e. if you can call the role because you're an architect (i.e. not root, or even an architect in a higher domain) then we check that you are calling the function in a child domain.

Reply:
Ah okay I see the different bitwise operation going on in `canCallBecause`. Maybe better to rename that function or even better, merge it into the `canCallBecauseArchitect` and rename _that_ one to something like `canCallBecauseOnlyArchitect` or something better?

Reply:
I think the name is ok -- `canCallBecause` tells you if someone can call a function _because_ they have a particular role (i.e. they aren't getting the permission elsewhere). I also think its cleaner to keep the `canCallBecause` in `DomainRoles` since it's accessing the underlying mappings and doing similar bit operations as the other functions there, and to keep it general since there's no reason to hardcode a particular role into `DomainRoles`. But an expository comment would be appropriate since it's not clear why we need to be doing this check at all for architect.

Reply:
I feel pretty strongly about this one actually and would want something renamed. I don't feel adding a comment will suffice. `canCallBecause` only half means your definition 

>someone can call a function because they have a particular role

The last part about the role is not obvious in that name. 
Plus it is only ever called once. I feel there are way too many functions in `ColonyStorage` serving `authDomain` and I understand some maybe justifiable because of the stack depth. But the following are not called anywhere else but that modifier and I'd argue for merging some of that logic:
- `isAuthorized`
- `validateDomainProof`
- `getChildSkillId` - that especially feel that should go
- `canCallBecauseArchitect` and `canCallBecause`

Reply:
Mm, those are reasonable points. I split them up that way largely to keep each function clean -- putting them all together will entail a big `if/else` statement and wont be as nice. I do think it's important to avoid creating local variables as we don't want to create problems for people writing functions which use these modifiers. But I also recognize that having functions which aren't used more than once creates its own confusion. I'll see if I can roll some of them together. I do think that `canCallBecause` belongs in `DomainRoles` though.

Reply:
I don't disagree with keeping `canCallBecause` in `DomainRoles`. I just disagree with its name, maybe `canCallBecauseHasSingleRoleOf` or something nicer you can come up with.

Reply:
Fair

Reply:
So I'll update the name to `canCallOnlyBecauseArchitect`, hopefully that's more helpful. With regards to the consolidating functions, I'll point out that this will involve _more_ lines of code. Consider:

```
  function validateDomainProof(uint256 permissionDomainId, uint256 childSkillIndex, uint256 childDomainId) internal view returns (bool) {
    if (permissionDomainId == childDomainId) {
      return true;
    } else {
      uint256 childSkillId = IColonyNetwork(colonyNetworkAddress).getChildSkillId(domains[permissionDomainId].skillId, childSkillIndex);
      return childSkillId == domains[childDomainId].skillId;
    }
  }
```

vs. the original

```
   function validateDomainProof(uint256 permissionDomainId, uint256 childSkillIndex, uint256 childDomainId) internal view returns (bool) {
    return (permissionDomainId == childDomainId) || (getChildSkillId(permissionDomainId, childSkillIndex) == domains[childDomainId].skillId);
  }

  function getChildSkillId(uint256 permissionDomainId, uint256 childSkillIndex) internal view returns (uint256) {
    return IColonyNetwork(colonyNetworkAddress).getChildSkillId(domains[permissionDomainId].skillId, childSkillIndex);
   }
```

8 lines vs 6, so two extra lines for the privilege of combining functions. I'll do it if you _really_ want but I don't think having the extra functions are a big deal and IMO they make this section easier to read.

Reply:
On the point of merging `getChildSkillId` in `validateDomainProof`, I was thinking of this end result:
In `authDomain`:
```
if (permissionDomainId != childDomainId) {
      require(validateDomainProof(_permissionDomainId, _childSkillIndex, _childDomainId), "ds-auth-invalid-domain-proof");
    }
```

as really the `validateDomainProof` check is only applicable to when we need to prove the child-parent relationship. 

```
function validateDomainProof(uint256 permissionDomainId, uint256 childSkillIndex, uint256 childDomainId) private view returns (bool) {
    uint256 childSkillId = IColonyNetwork(colonyNetworkAddress).getChildSkillId(domains[permissionDomainId].skillId, childSkillIndex);
    return  (childSkillId == domains[childDomainId].skillId);
  }
```
Which is 6 lines of code.

Also a bit on the naming (sorry I think I left a note on that somewhere but maybe didn't? As in that "proof" is used in Merkle tree proofs in reputation mining and shouldn't be used here) `validateDomainProof` -> `validateDomainInheritance` (Open to better names here)

`isAuthorized` I'm okay to keep separate, as it's reused in 2 places and as you say matches the `DSAuth` design.

For `canCallBecauseArchitect`, I prefer your suggestion of `canCallOnlyBecauseArchitect`.

Reply:
Mm that's how I did the domain proof the first time around, but I combined it into a single function on the grounds that those two pieces of logic always go together, plus I liked the idea of `authDomain` being mostly just one-line requires. But I can change it back.

And fair, I can reserve the term `proof` for Merkle proofs (although this is a ~kind of proof).